### ISSUE TITLE
Support sharing certificate over multiple configuration files

### ISSUE THREAD

**Issue by JonasAlfredsson** [JonasAlfredsson/docker-nginx-certbot#91]
Right now there is a limitation that every `server` that uses the same `cert_name` for their keys need to have them all defined in the same config file, since otherwise we will overwrite it in a following request. 

### Example:

```
server {
    listen              443 ssl;
    server_name         yourdomain.org www.yourdomain.org;
    ssl_certificate_key /etc/letsencrypt/live/test-name/privkey.pem;
    ...
}

server {
    listen              443 ssl;
    server_name         sub.yourdomain.org;
    ssl_certificate_key /etc/letsencrypt/live/test-name/privkey.pem;
    ...
}
```

These two servers will create a single certificate file (`/etc/letsencrypt/live/test-name/privkey.pem`) which are valid for the following domains:

- `yourdomain.org `
- `www.yourdomain.org`
- `sub.yourdomain.org`

However, if these were split across two different config files we would first create a certificate that was valid for 

- `yourdomain.org `
- `www.yourdomain.org`

and then overwrite it with a certificate that was only valid for 

- `sub.yourdomain.org`

This is not optimal, and a solution could be to iterate over all configuration files and build a hashmap/[associative array](https://www.artificialworlds.net/blog/2012/10/17/bash-associative-array-examples/) with all the domains associated with a specific certificate name and then combine them all to a single request at the end. 

This is not optimal in Bash, since it would be really beneficial to have the value of the keys to be arrays but that is not supported. A solution would then be to create something like this:

```bash
[${cert_name}] = "domain_name,domain_name"
```

where we then use `cut -d,` or similar to extract the domains afterwards. The benefit to this is that we can give the user much more freedom to how they want to structure their configuration files, and perhaps reduce the amount of requests we have to make towards Let's Encrypt.

**Comment by maxkagamine** [comment]
I think I've actually come up with a good solution for this, taking advantage of the fact that a) the script will symlink but otherwise ignore confs that don't have an `ssl_certificate`, and b) nginx treats wildcards as a catch-all, so a conf with the exact name will match first and override it ([docs](http://nginx.org/en/docs/http/server_names.html)):

**user_conf.d/default.conf**

```nginx
server {
  # Listen to port 443 on both IPv4 and IPv6.
  listen 443 ssl default_server reuseport;
  listen [::]:443 ssl default_server reuseport;

  # Domain names this server should respond to.
  server_name example.com *.example.com;

  # Load the certificate files.
  ssl_certificate         /etc/letsencrypt/live/default/fullchain.pem;
  ssl_certificate_key     /etc/letsencrypt/live/default/privkey.pem;
  ssl_trusted_certificate /etc/letsencrypt/live/default/chain.pem;

  # Load the Diffie-Hellman parameter.
  ssl_dhparam /etc/letsencrypt/dhparams/dhparam.pem;

  deny all;  
}
```

**includes/ssl**

```nginx
listen 443;

# Load the certificate files.
ssl_certificate         /etc/letsencrypt/live/default/fullchain.pem;
ssl_certificate_key     /etc/letsencrypt/live/default/privkey.pem;
ssl_trusted_certificate /etc/letsencrypt/live/default/chain.pem;

# Load the Diffie-Hellman parameter.
ssl_dhparam /etc/letsencrypt/dhparams/dhparam.pem;
```

**user_conf.d/foo.conf**

```nginx
server {
  include includes/ssl;

  server_name foo.example.com;

  return 200 "Hello";
  add_header Content-Type text/plain;
}
```

**Comment by XaF** [comment]
I was planning to take a look at that actually, as at the moment there was also the fact that if you add another `server` block in the same file, it is picked up and tries to add the `server_name` in the cert, which could probably be handled better.

As per multi-values associative arrays, the comma-separated approach works well. I've also seen `array["key,IDX"]` in the past but that's not as performant.

I believe this can be done in bash still, by reading line by line though. Should still be fast enough!

**Comment by XaF** [comment]
Actually smart @maxkagamine!
This is true that the include won't be followed, and the deny all in your wildcard domain solves the problem. Pretty neat approach to get around it here :)

**Comment by JonasAlfredsson** [comment]
Neat little workaround there, however, would not Nginx panic when we start this image?
1. Symlink all .conf files to `/etc/nginx/user_conf.d/`
2. Disable all which have missing files (this includes `default.conf` but not `foo.conf`)
3. Nginx starts and loads `foo.conf` which includes the certificates that does not exist yet

**Comment by maxkagamine** [comment]
Hm, good point, the certs would need to be created first.  
For `example.com`, it could be in its own conf file,  but to keep it in the same cert you could do:

```nginx
if ($host != example.com) {
  return 444;
}

return 200 "Hello from example.com";
add_header Content-Type text/plain;
```

which seems to do the job, though `ssl_reject_handshake on;` can't be used at that point.

**Comment by maxkagamine** [comment]
I opened a pr with some docs on the current approach I'm using. Let me know what you think.

**Comment by JonasAlfredsson** [comment]
One of the main points when I chose this image to continue to build upon was that it was able to start completely by itself (by automatically renaming files), so I would not like to move away from that really nice feature if possible. 
Thus I will not accept the currently suggested solution, even though it will suffice as a neat little workaround until a more solid one is created :)

**Comment by XaF** [comment]
Just created #94 that should solve this :)

**Comment by maxkagamine** [comment]
@XaF Wasn't expecting this to be implemented so quickly😅 You are awesome.

**Comment by JonasAlfredsson** [comment]
This should now be included in the [v3.0.0](https://github.com/JonasAlfredsson/docker-nginx-certbot/releases/tag/v3.0.0) release.
